### ISSUE TITLE
crisprfinder script not working for some examples

### ISSUE THREAD

**Issue by mbonsma** [phageParser/phageParser#239]
I'm trying to use [`crisprfinder.py`](https://github.com/phageParser/phageParser/blob/master/parserscripts/crisprfinder.py) on a big file, but it returns no output even though running the same file manually on the [website](http://crispr.i2bc.paris-saclay.fr/cgi-bin/crispr/advSavefiles.cgi?FN=NC_011091.fasta&DIR=138.51.240.91_Aug_15_2017_18_50_44) seems to work fine. 

[This bacteria genome](https://github.com/phageParser/phageParser/blob/master/data/NC_000853.fasta) works fine using the script:

```
python parserscripts/crisprfinder.py data/NC_000853.fasta out.txt
```

But when I tried to run a different genome (NC_011091, it's a fruit fly, don't worry about it), it opened the browser, then closed after about 3 seconds with an empty output file. The same file uploaded after about 20 seconds on the website and found some possible CRISPRs. 

They're both fasta files. The only thing I can imagine is that the script isn't waiting for the big file to finish uploading? I don't know enough about selenium / web scraping to know what to do here.

**Comment by mbonsma** [comment]
Update: the CRISPRdb website has a 67 million base limit for their system supposedly (although this doesn't apply if you upload your own file?), so we can split genomes into separate files for each chromosome and loop through them that way.

**Comment by thisisjaid** [comment]
hi @mbonsma. I've been toying with this all day today as a tiny project and I have to say there's a lot of things broken with http://crispr.i2bc.paris-saclay.fr. I don't think the failure you're seeing in this script has to do with the number of base pairs you're allowed to submit to be honest as the file upload through the browser works fine. The clue I think is in the really long time it takes to process the file when you upload through the browser. 

From the tests I did both with curl and python requests it's pretty obvious that the CGI script is somehow returning some of the content to the HTTP request but not all of it and the crucial parts with the actual search results are effectively missing even though the file uploads fine and the header parts of the fasta file are displayed in the HTTP return correctly.

I've tried randomly slicing the file into 200k lines and it seems to work fine with that much. I presume from an analysis pov it's a better idea to slice it up into chromosome and I'd like to take a shot at implementing this, could you let me know how I'd go about doing that split as the fasta file seems like a continuous stream with no separators at the moment.

Also, would you mind me trying to re-write this script to use requests/beautifulsoup instead of selenium as that seems very complicated and getting the raw HTML via requests worked just fine in my tests.

**Comment by mbonsma** [comment]
Hi @thisisjaid, that'd be awesome! 

1. Splitting by chromosome: as far as my limited knowledge goes, the accession numbers we've been using to download genomes return them in chromosome chunks --- each new chromosome would have a fasta header like this `> some title info blah blah`. Which fasta file were you using to test this out? 

2. It would be great if you re-wrote it - we have no reason to use selenium over anything else, and I've played around with beautiful soup in the past.

**Comment by thisisjaid** [comment]
Ah I see, I didn't realize when I grabbed the fasta file that it was actually for a single chromosome. The 'whole genome shotgun sequence' threw me off and I thought that was a full genome sequence. I was running it on NT_167065.1 which is chromosome 3R of the same fruitfly you were testing with.

So in that case we still have a problem as the fasta file I was running _was_ a single chromosome and it still failed. Cutting the file in half made it work, but then that would impede the analysis. I'm still somewhat stumped as to why this is occurring and the different behaviour between browser and headless requests. Few more tests and it seems 350.000 lines are ok but 375.000 is not, so somewhere in that region the problem occurs.

**Comment by mbonsma** [comment]
That is strange indeed.

It wouldn't be the worst thing to split chromosomes into smaller chunks. After running through CRISPRfinder, we could check for any hits that were close to the cut point and re-do those with a different split site to be sure. I expect there wouldn't be very many to re-run - genomes are big and CRISPR loci are small :) 

Also, we've been in touch with the makers of CRISPRfinder and they sent us the perl script. It doesn't give us the same results as the website, but if you like I can send you the script anyway if you want to poke around. I'm hesitant to post it here, but I can email it to you.

**Comment by thisisjaid** [comment]
Yeah I was thinking along the same lines. We can split progressively ie: lines 0-2000 1000-3000 etc, parse the results on each run, dump them into a pandas table and then just drop duplicate rows at the end and it should do the job.

I am curious however to have a look at the perl script so if you can foward that to thisisjaid@gmail.com I'll have a look and see if we can use it.

**Comment by mbonsma** [comment]
@thisisjaid was able to get the local version of CRISPRfinder working, but I think it's still a good idea to have this as an option if you're willing to work on it. So yes, please do submit a PR!
